### PR TITLE
Add geo types sync integration tests.

### PR DESCRIPTION
**Pull request by clementetb**
Add tests to validate we can define sync subscriptions using geo queries.

### CODE REVIEW

File: packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/internal/UserImpl.kt [hunk 1]
```diff
@@ -49,7 +49,7 @@ public class UserImpl(
     override val loggedIn: Boolean
         get() = RealmInterop.realm_user_is_logged_in(nativePointer)
     override val provider: AuthenticationProvider
-        get() = AuthenticationProviderImpl.fromId(RealmInterop.realm_user_get_auth_provider(nativePointer))
+        get() = identities.first().provider
```
Comment (cmelchior) [inline packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/internal/UserImpl.kt:52]: Talking to JS, we should deprecate this method as it doesn't really make sense anymore and instead point people towards `user.identities`.

File: packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+import kotlinx.coroutines.delay
+import org.mongodb.kbson.ObjectId
+import kotlin.test.AfterTest
+import kotlin.test.BeforeTest
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertFails
+import kotlin.time.Duration.Companion.seconds
+
+private val GEO_SCHEMA = setOf(SyncRestaurant::class, Location::class)
```
Comment (cmelchior) [inline packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt:33]: We cannot use this anymore. All schemas need to be the same after the server introduced the new developer mode. I have a branch you can work from.

Comment (cmelchior) [reply packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt:33]: See https://github.com/realm/realm-kotlin/pull/1535

File: packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+@OptIn(ExperimentalGeoSpatialApi::class)
+class GeoSpatialTests {
+    private lateinit var app: TestApp
+
+    @BeforeTest
+    fun setup() {
+        app = TestApp(this::class.simpleName, appName = TEST_APP_FLEX, logLevel = LogLevel.ALL)
```
Comment (cmelchior) [inline packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt:41]: LogLevel.ALL seems a bit much?

File: packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        val config =
+            SyncConfiguration.Builder(
+                user = user,
+                schema = GEO_SCHEMA
+            ).initialSubscriptions(rerunOnOpen = true) {
+            }.build()
+
+        Realm.open(config).use {
+            it.writeBlocking {
+                assertFails {
```
Comment (cmelchior) [inline packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt:91]: I don't think the write will fail as long as the class is in the schema, instead you will probably see a compensating write?

File: packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    ) {
+        val section = ObjectId()
+
+        // User #1 will try to write some data and assert some failure conditions.
+        val user1 = createRandomUser()
+        val config =
+            SyncConfiguration.Builder(
+                user = user1,
+                schema = GEO_SCHEMA
+            ).initialSubscriptions(rerunOnOpen = true) {
```
Comment (cmelchior) [inline packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt:160]: Why rerun on open?

File: packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+                        "section = $0 AND location GEOWITHIN $1",
+                        section,
+                        bounds
+                    )
+                )
+            }.waitForInitialRemoteData().build()
+
+        Realm.open(config).use { realm ->
+            runBlocking {
+                realm.subscriptions.waitForSynchronization(timeout = 30.seconds)
```
Comment (cmelchior) [inline packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt:172]: This shouldn't be needed if you have `waitForInitialRemoteData`. It should take into account initial subscriptions.

File: packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+                    SyncRestaurant().apply {
+                        this.section = section
+                        this.location = validLocation
+                    }
+                )
+            }
+
+            runBlocking {
+                realm.syncSession.uploadAllLocalChanges(timeout = 30.seconds)
+                delay(30000)
```
Comment (cmelchior) [inline packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt:194]: Why this?

File: packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+                    it.query<SyncRestaurant>(
+                        "section = $0 AND location GEOWITHIN $1",
+                        section,
+                        bounds
+                    )
+                )
+            }.waitForInitialRemoteData().build()
+
+        Realm.open(config).use { realm ->
+            runBlocking {
```
Comment (cmelchior) [inline packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt:171]: There are a lot of runBlocking calls in here. I would probably just move it outside the `Realm.open()` and then cleanup the code here.

File: packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt (hunk 4 above)
Comment (clementetb) [reply packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt:91]: It fails with `RLM_ERR_NO_SUBSCRIPTION_FOR_WRITE`. It is core preemptively catching the error.

File: packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt (hunk 7 above)
Comment (clementetb) [reply packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt:194]: I forgot to remove it. I wanted to check why the data is deleted after setting an invalid location on a polygon shape, while Circle and Box undo the changes.

File: packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt (hunk 5 above)
Comment (clementetb) [reply packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/common/GeoSpatialTests.kt:160]: Initially, I reused the Realms and then I did not remove this line.

**Review by cmelchior — APPROVED** [review]
:shipit:

**Pull request merged**